go/vt/topo/stats_conn.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2018 The Vitess Authors
+ Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+
+	"vitess.io/vitess/go/stats"
+)
+
+var _ Conn = (*StatsConn)(nil)
+
+var (
+	topoStatsConnTimings = stats.NewMultiTimings(
+		"TopologyConn",
+		"TopologyConn timings",
+		[]string{"Operation", "Cell"})
+	topoStatsConnCounts = stats.NewCountersWithMultiLabels(
+		"TopologyConnCounts",
+		"Operation counts to topology cells",
+		[]string{"Operation", "Cell"})
+
+	topoStatsConnErrors = stats.NewCountersWithMultiLabels(
+		"TopologyConnErrors",
+		"Operation errors to topology cells",
+		[]string{"Operation", "Cell"})
+)
+
+// The StatsConn is a wrapper for a Conn that emits stats for every operation
+type StatsConn struct {
+	cell string
+	conn Conn
+}
+
+// NewStatsConn returns a StatsConn
+func NewStatsConn(cell string, conn Conn) *StatsConn {
+	return &StatsConn{
+		cell: cell,
+		conn: conn,
+	}
+}
+
+// ListDir is part of the Conn interface
+func (st *StatsConn) ListDir(ctx context.Context, dirPath string, full bool) ([]DirEntry, error) {
+	startTime := time.Now()
+	statsKey := []string{"ListDir", st.cell}
+	defer topoStatsConnTimings.Record(statsKey, startTime)
+	res, err := st.conn.ListDir(ctx, dirPath, full)
```

Comment:
I'm not sure how much visibility you want into the class of failures but it potentially useful to tag based on the error code. So something like adding:

```go
// vt/topo/errors.go
// caveat: ideally we'd convert this back to *string for display or have a parallel ErrorCodeName
func ErrorCode(e error) *ErrorCode {
	if e == nil { return nil }
	switch err := e.(type) {
	case *Error:
		return err.code
	case Error:
		return err.code
	default:
		return nil
	}
}
```

and then in your error handling (I'm taking some liberty with go for brevity :heart: ... and also eliding some of the necessary error handling / details)

```go
if err != nil {
  topoStatsConnErrors.Add(statsKey, int64(1))
  topoStatsConnScopedErrors.Add([...statsKey, topo.ErrorCode(err)], int64(1))
}
```

Reply:
I think for now I would like to keep the generic error count. I'm mostly interested in tracking general errors as we make the changes from skipping consul agent and talking to consul via the HA proxy vs normal operations. 

Should we iterate in the future if we see need for more granularity ?

Reply:
Minor, not strong, objections; presumably something downstream of these calls is logging error codes so we can check aggregation for details if necessary?

My thought here is that if our error counts are non-zero (and we don't know because we don't have telemetry right now, correct?) then moving from non-zero to non-zero as we shift between consul and consul-via-proxy doesn't actually tell us if we've functionally changed the behavior since the type of errors are also important.

Two caveats

1. If we know/suspect there are no errors in steady-state then this is less important
2. If my read of `stats.CountersWithMultiLabels` is incorrect and it's more effort than just adding a single scoped counter then the effort:(payoff | code noise) ratio changes and it's less compelling for me